File: 96-unique-binary-search-trees/unique-binary-search-trees.cpp
class Solution {
public:
    int numTrees(int n) 
    {
        vector<int> dp(n + 1, 0);
        dp[0] = dp[1] = 1; 

        // Bottom-up DP computation
        for(int node = 2; node <= n; node++) 
        {
            for(int root = 0; root < node; root++) 
            dp[node] += dp[root] * dp[node - 1 - root];
        }

        return dp[n]; 
    }
};

File: 96-unique-binary-search-trees/unique-binary-search-trees.java
class Solution {
    private Integer[] dp;
    private int solve(int n){
        if(n == 0) return 1;

        if(dp[n] != null) return dp[n];

        int count = 0;
        for(int i = 1; i <= n; i++){
            count += (solve(i - 1) * solve(n - i));
        }
        return dp[n] = count;
    }
    public int numTrees(int n) {
        dp = new Integer[n + 1];
        return solve(n);
    }
}